ewatercycle/forcing.py

```diff
@@ -0,0 +1,341 @@
+    var_names = 'tas', 'pr', 'psl', 'rsds', 'rsdt'
+
+    if startyear is not None:
+        for var_name in var_names:
+            variables[var_name]['start_year'] = startyear
+
+    if endyear is not None:
+        for var_name in var_names:
+            variables[var_name]['end_year'] = endyear
+
+    return data
+
+
+def update_pcrglobwb(data: dict,
+                     *,
+                     forcings: list = None,
+                     basin: str = None,
+                     startyear: int = None,
+                     endyear: int = None,
+                     startyear_climatology: int = None,
+                     endyear_climatology: int = None,
+                     extract_region: dict = None):
+    """
+    Update pcrglobwb recipe data in-place.
+    
+    Parameters
+    ----------
+    forcings : list
+        List of forcings to use
+    basin : str
+        Name of the basin (used for data output filename only)
+    startyear : int
+        Start year for the observation data
+    endyear : int
+        End year for the observation data
+    startyear_climatology : int
+        Start year for the climatology data
+    endyear_climatology : int
+        End year for the climatology data
+    extract_region : dict
+        Region specification, must contain `start_longitude`, 
+        `end_longitude`, `start_latitude`, `end_latitude`
+    """
+    preproc_names = ('crop_basin', 'preproc_pr', 'preproc_tas',
+                     'preproc_pr_clim', 'preproc_tas_clim')
+
+    if forcings is not None:
+        datasets = [FORCINGS[forcing] for forcing in forcings]
+        data['diagnostics']['diagnostic_daily'][
+            'additional_datasets'] = datasets
+
+    if basin is not None:
+        data['diagnostics']['diagnostic_daily']['scripts']['script'][
+            'basin'] = basin
+
+    if extract_region is not None:
+        for preproc_name in preproc_names:
+            data['preprocessors'][preproc_name][
+                'extract_region'] = extract_region
+
+    variables = data['diagnostics']['diagnostic_daily']['variables']
+    var_names = 'tas', 'pr'
+
+    if startyear is not None:
+        for var_name in var_names:
+            variables[var_name]['start_year'] = startyear
+
+    if endyear is not None:
+        for var_name in var_names:
+            variables[var_name]['end_year'] = endyear
+
+    var_names_climatology = 'pr_climatology', 'tas_climatology'
+
+    if startyear_climatology is not None:
+        for var_name in var_names_climatology:
+            variables[var_name]['start_year'] = startyear_climatology
+
+    if endyear_climatology is not None:
+        for var_name in var_names_climatology:
+            variables[var_name]['end_year'] = endyear_climatology
+
+    return data
+
+
+def update_wflow(
+    data: dict,
+    *,
+    forcings: list = None,
+    startyear: int = None,
+    endyear: int = None,
+    extract_region: dict = None,
+    dem_file: str = None,
```

Comment:
In a far future we might want to make it possible to derive the DEM file directly from a shapefile. For now I think it's unavoidable that we have a deviating call signature here.

Reply:
All function calls are slightly different for that matter, we will get back to this at a later point 👍